app/src/main/java/com/epfl/drawyourpath/database/FireDatabase.kt

```diff
@@ -38,76 +70,175 @@ class FireDatabase : Database() {
+                        userAccount.put("username", username)
+                        database.child("users").child(userId).updateChildren(userAccount as Map<String, Any>)
+                            .addOnSuccessListener { future.complete(true) }
+                            .addOnFailureListener {
+                                future.completeExceptionally(Exception("Impossible to create the user account."))
+                            }
+                    }
+                    .addOnFailureListener {
+                        future.completeExceptionally(java.lang.Error("Impossible to find the link between the username and the userId to the Database.") )
+                    }
+            }else{
+                future.completeExceptionally(java.lang.Error("The username is not available !"))
+            }
+        }
+        return future
     }
 
-    override fun setUsername(username: String) {
-        val userAdd = HashMap<String, String>()
-        userAdd.put(username, "empty")
-        database.child("users").updateChildren(userAdd as Map<String, Any>)
-            .addOnFailureListener { throw Exception("Impossible to add the username on the database") }
+    override fun initUserProfile(userModel: UserModel): CompletableFuture<Boolean> {
+        val userData = HashMap<String, Any>()
+        userData.put("email", userModel.getEmailAddress())
+        userData.put("firstname", userModel.getFirstname())
+        userData.put("surname", userModel.getSurname())
+        userData.put("dateOfBirth", userModel.getDateOfBirth().toEpochDay())
+        userData.put("distanceGoal", userModel.getDistanceGoal())
+        userData.put("activityTimeGoal", userModel.getActivityTime())
+        userData.put("nbOfPathsGoal", userModel.getNumberOfPathsGoal())
+
+        val future = CompletableFuture<Boolean>()
+        val userId = getUserId()
+
+        database.child(usersProfileFileName).child(userId).updateChildren(userData)
+            .addOnSuccessListener { future.complete(true) }
+            .addOnFailureListener { future.completeExceptionally(java.lang.Error("Impossible to init the user data on the database.")) }
+        return future
     }
 
-    override fun setPersonalInfo(
-        username: String,
-        firstname: String,
-        surname: String,
-        dateOfBirth: LocalDate
-    ) {
-        val userAdd = HashMap<String, String>()
-        userAdd.put("firstname", firstname)
-        userAdd.put("surname", surname)
-        updateUserData(userAdd, username)
-
-        //for the date
-        val unixDate: Long = dateOfBirth.toEpochDay()
-        val userDateMap = HashMap<String, Long>()
-        userDateMap.put("dateOfBirth", unixDate)
-        updateUserData(userDateMap, username)
+    override fun getUserAccount(userId: String): CompletableFuture<UserModel> {
+        val future = CompletableFuture<UserModel>()
+
+        database.child(usersProfileFileName).child(userId).get().addOnSuccessListener { userData ->
```

Comment:
I think there is a problem with how data is stored. Here we fetch all the data present at `users/{uid}/`. We should break down the data in the database in smalled chunks. For example, we could create a `users/{uid}/public` node to only fetch profile information to init the view model. It would prevent fetching the entire data of the user.

Reply:
I think we could do that in future task when we will choose the privacy rules

Reply:
It isn't really linked to privacy, as now you would fetch all the races, etc. of the user just to get basic data.

Reply:
At this time all the information store in the userModel are useful. I don't really understand what you want to do.